기본적인 프론트 코드 작성

BuchuKim opened the pull request:
## 개요
프론트 코드와 프론트가 추가되면서 필요한 기능 및 엔드포인트를 쫌쫌따리 추가했습니다..

## 작업사항
- script로 쿼리를 날리는데 xml, ajax 등의 방법을 고려하다 `axios`를 이용했습니다. 간단한 요청 구성이 마음에 들어서 사용했는데, 따로 모듈을 import하거나 cdn을 받아와야 한다는게 단점이라면 단점이겠네요?
- 인덱스 페이지에 띄울 추천 레시피 기능을 만들었습니다... 리포지토리의 `getFeatured()`를 통해 `recipeQueryService.getTopRecipe()`가 호출되는 구조입니다.. 쿼리서비스의 script를 사용했어욥.
- 세션때 말씀하신 태그 목록을 따로 서버에서 불러오도록 했고, 로그인 체크 엔드포인트도 추가했습니다. 로그인 체크는 로그인 성공 시에 세션에 존재할 `SessionUser`의 존재여부로 판단했고 인증이 안되어있을 경우 401에러를 던지도록 했습니닷
- `ViewController`를 통해 일단 static html 파일을 반환할 컨트롤러를 만들었습니다
- es document 저장 과정에서.. 저번 커밋에서 ingredient 추가한게 누락돼서 여기 커밋에 부랴부랴 넣을게요....
- 나름 반복되는건 함수화 시키거나 js 파일을 따로 분리했는데 코드가 많이 중구난방이네요ㅜㅜ

## preview
- `index.html` : 인덱스 페이지입니다. featured(추천) 레시피가 메인에 뜨도록 구성했어요 <img width="1470" alt="image" src="https://github.com/f-lab-edu/recipetory/assets/69233405/76b2b6c4-3a88-423e-8443-fd4fa459dd31">
- `tag-search.html` : 태그 검색 페이지입니다. 여러 개의 태그를 선택할 수 있고, 검색하기 버튼을 누르면 선택된 태그가 모두 포함된 레시피 목록을 아래 띄워주는 구조입니다 <img width="1470" alt="image" src="https://github.com/f-lab-edu/recipetory/assets/69233405/991d90b9-0cd0-4371-9f4b-b8e31dad3c71">
- `recipe-new.html` : 레시피 생성 폼입니다. 세션때 보여드린 것과 같은 구조에요. <img width="1470" alt="image" src="https://github.com/f-lab-edu/recipetory/assets/69233405/54a91dcd-dec4-4aa0-9c74-7738f0c64105">
- `recipe-search.html` : 상단 검색창에서 할 수 있는 키워드 검색 결과창입니다. <img width="1470" alt="image" src="https://github.com/f-lab-edu/recipetory/assets/69233405/7fea6ebc-3372-4305-9b57-01d085f291c8">
- `recipe-show.html` : 레시피 조회 뷰인데, 카드 아래의 버튼을 통해 댓글 및 리뷰 조회/추가가 가능합니다 <img width="1470" alt="image" src="https://github.com/f-lab-edu/recipetory/assets/69233405/10cd399e-795f-4617-8e6c-32c04b9dbe8c"> 리뷰/댓글 작성을 클릭하면 모달창으로 작성 폼이 뜰 수 있도록 했어요 <img width="1470" alt="image" src="https://github.com/f-lab-edu/recipetory/assets/69233405/e424d699-ed87-4f9d-a9cd-a25a6c0db0a1">

BuchuKim committed:
feat : auth check endpoint 추가
- spring security가 기본적으로 제공하는 redirect 대신 401 에러를 날리도록 구성했습니다
- security filter chain 통일

BuchuKim committed:
feat : feature recipe 기능
- 조회수와 평점을 기준으로 상위 레시피를 뽑아 추천탭에 넣는 로직 추가
- es script 사용했습니다

BuchuKim committed:
fix : recipe document에서 ingredient 누락 수정
- session user에 userId 포함하도록 추가

BuchuKim committed:
feat (#23) : 프론트 페이지 작성
- index, 레시피 생성폼, 레시피 조회, 이름 검색, 태그 검색, 리뷰/댓글 폼 및 목록 보기까지 구현

BuchuKim committed:
feat : tag 목록을 반환하는 end point, dto 추가

BuchuKim committed:
fix : featured test todo

BuchuKim commented:
[http://34.64.192.158](http://34.64.192.158)
데이터는 오늘 중으로 쌓아볼건데, 일단 사이트 어떻게 생겼는지 직접 보는게 나을 것 같아서 배포 서버도 추가해욥.

f-lab-fi commented on src/main/java/com/recipetory/recipe/infrastructure/RecipeDocumentQueryService.java:
```diff
@@ -81,4 +82,31 @@ public List<RecipeDocument> findByRecipeInfo(
         }
     }
+
+    /**
+     * view count, ratings를 기준으로 레시피를 내림차순으로 정렬하여
+     * 인자로 전달된 갯수만큼의 레시피를 반환한다.
+     * @param number
+     * @return recipe document
+     */
+    public List<RecipeDocument> getTopRecipe(int number) {
+        try {
+            // 평점(ratings), 조회수(viewCount)를 이용해 추천에 사용될 점수를 계산하는 script
+            String script = "doc['recipeStatistics.ratings'].value + " +
+                    "Math.log(1 + doc['recipeStatistics.viewCount'].value);";
+
+            // nested field의 값을 이용한 결과를 내림차순으로 정렬하는 쿼리 source
+            ScriptSort source = SortOptionsBuilders.script()
+                    .script(Script.of(sc -> sc.inline(InlineScript.of(i -> i.source(script)))))
+                    .type(ScriptSortType.Number)
+                    .order(SortOrder.Desc)
+                    .nested(NestedSortValue.of(sv -> sv.path("recipeStatistics"))).build();
+
+            SearchResponse<RecipeDocument> response = esClient.search(
+                    s -> s.index("recipe").size(number).sort(so -> so.script(source)), RecipeDocument.class);
+            return response.hits().hits().stream().map(Hit::source).toList();
+        } catch (IOException e) {
+            throw new EsClientException("recipe");
+        }
+    }
```
굳ㅋ

f-lab-fi commented on src/main/resources/static/index.html:
```diff
@@ -0,0 +1,104 @@
+<!DOCTYPE html>
+<html lang="en">
```
lang="en" 는 무슨 속성일까요~?

BuchuKim replied:
HTML에서 기본적으로 사용하는 언어 속성을 지정하는 태그입니다..
웹 접근성과 관련한 내용으로, HTML을 읽는 브라우저의 언어 특정적인 기능(ex. 번역, 자동읽기 등) 실행에 사용됩니다. `lang="en"`은 현재 HTML 파일의 기본 언어가 en, 즉 english라고 명시한 것이고 따라서 브라우저는 해당 파일이 파일이 영어로 작성되어있다고 가정하고 관련 기능을 수행합니다.

스프링 시큐리티의 oauth client에서 제공하는 기본 로그인창을 복붙하면서 살을 붙였는데 html 파일 전부가 en으로 되어있네요.... 수정하겠습니다 ㅜㅅㅜ

f-lab-fi commented on src/main/resources/static/js/build-menu.js:
오호 좋은데요? ㅎ
footer 쪽 코멘트 남기려다가 말았는데, footer도 지금 하신거랑 비슷하게 처리 가능합니다~
약간 잡기술?

```html
<div id="navbar"></div>
<script>
  fetch("navbar.html")
    .then(response => response.text())
    .then(data => {
      document.getElementById("navbar").innerHTML = data;
    });
</script>
```

f-lab-fi commented on src/main/resources/static/user-show.html:
```diff
@@ -0,0 +1,156 @@
+    <script src="/view/js/build-menu.js"></script>
+    <script>
+      const createUserCard = (user) => {
+        const card = document.createElement("div");
+        card.classList.add("card", "mb-3");
+
+        const cardBody = document.createElement("div");
+        cardBody.classList.add("card-body");
+
+        const name = document.createElement("h5");
+        name.textContent = user.name;
+        name.classList.add("card-title");
+
+        const bio = document.createElement("p");
+        bio.classList.add("card-text");
+        bio.textContent = user.bio;
+
+        cardBody.appendChild(name);
+        cardBody.appendChild(bio);
+
+        card.appendChild(cardBody);
+
+        return card;
```
저는 HTML 요소 추가하는건 아래 링크에서 쓴 방식이 좋은 것 같아요 ㅎ

https://github.com/f-lab-edu/recipetory/pull/34/files#diff-92c6d8ccfaf34539da61b4ac4b0a70534445d220c14e8b6361b486fa587a27baR115-R126

지금 코멘트 드린 방식은 결과적으로 어떤 태그가 만들어지는지 가시적으로 안보여서 저는 별로 선호 안합니다. :)

BuchuKim replied:
오히려 링크 부분이 html로서 가시성은 좋지만 "이럴거면 쌩 HTML을 짜지 뭣하러 이렇게 구성하냐." 싶을 것 같았는데 .. 
멘토님 말씀 듣고 다시 생각해보니 눈에 보여지는 페이지 작성하는데 가시성보다 중요한게 뭐가 있겠나 싶네욥ㅋㅋ 쫌쫌따리 고쳐보겠슴다..

f-lab-fi commented:
@BuchuKim 

수고하셨습니당 ㅎ
http://34.64.192.158/ 여긴 저는 접속이 안되는듯?! 방화벽인지..

> script로 쿼리를 날리는데 xml, ajax 등의 방법을 고려하다 axios를 이용했습니다. 간단한 요청 구성이 마음에 들어서 사용했는데, 따로 모듈을 import하거나 cdn을 받아와야 한다는게 단점이라면 단점이겠네요?
-> ㅇㅈ 그냥 fetch 써도 될듯합니다. 근데 또 프론트엔드 개발자 입장에선 이미 구현되어있는 axios 불러와서 사용하는게 더 좋을 수도 있어요. 왜냐하면 fetch 같은걸 직접 사용하면 특정 웹 브라우저에서는 지원 안할수도 있습니다. 근데 보통 라이브러리 불러쓸 때는 해당 라이브러리에 여러 웹 브라우저에 대한 호환성이 제공되도록 코드가 작성되어 있을 수도 있으니까요.

> es document 저장 과정에서.. 저번 커밋에서 ingredient 추가한게 누락돼서 여기 커밋에 부랴부랴 넣을게요....
-> 네 ㅎ

> 나름 반복되는건 함수화 시키거나 js 파일을 따로 분리했는데 코드가 많이 중구난방이네요ㅜㅜ
-> 백엔드 개발자가 이정도면 됐죠 뭐 ㅎ 너무 신경 쓰지 마시죠~

BuchuKim committed:
feat (#23) : 프론트 페이지 추가 수정

BuchuKim committed:
fix : 북마크 목록의 레시피 자체를 반환하도록 bookmark 서비스 코드 수정
- 북마크 추가시 auth 검사할 수 있도록 security config 수정

BuchuKim committed:
feat : recipe document 연관관계에서 author을 갖도록 기능 추가

BuchuKim committed:
fix : ingredient validation 길이 수정

BuchuKim committed:
chore : 배포 관련 static, test 관련 security 설정파일 작성

BuchuKim committed:
fix : build tag의 tag 변수 초기화 오류 수정

BuchuKim committed:
fix : 프론트 코드 수정
- 404, 401, 500 에러에 대응될 html + 스크립트 파일
- 표시할 자원이 없을 때 존재하지 않음을 명시
- tags 검색에 현재 검색중인 태그가 추가되도록 구성

BuchuKim committed:
chore : recipe - bookmark cascade 수정

BuchuKim committed:
feat : 레시피 목록 개수 제한
- 상위 100개만 제한하도록 수정
